Event dispatching adapter

dantleech opened the pull request:
This PR an EventDispatchingAdapter which wraps the configured adapter and dispatches events, would be a replacement for: https://github.com/symfony-cmf/RoutingAuto/pull/23

dantleech commented on Tests/Unit/Adapter/EventDispatchingAdapterTest.php:
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Unit\Adapter;
```
Bad NS and missing header

dantleech commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ dev-master
                   instead of or in addition to URI strings.
 * **2015-01-29**: Added SymfonyContainerParameterTokenProvider for retrieving
                   path elements from the DI container
+* **2014-12-07**: Added event dispatcher and URI_CONTEXT_BUILT event
```
Wrong note

dantleech commented on AutoRouteManager.php:
```diff
@@ -75,11 +94,19 @@ public function buildUriContextCollection(UriContextCollection $uriContextCollec
+            if ($this->eventDispatcher) {
+                $event = new UriContextEvent($uriContext);
+                $this->eventDispatcher->dispatch(RoutingAutoEvents::URI_CONTEXT_BUILT, $event);
+            }
```
Remove this.

dantleech commented on Adapter/EventDispatchingAdapter.php:
```diff
@@ -0,0 +1,123 @@
+use Symfony\Cmf\Component\RoutingAuto\Event\AutoRouteCreateEvent;
+use Symfony\Cmf\Component\RoutingAuto\Event\AutoRouteMigrateEvent;
+
+class EventDispatchingAdapter implements AdapterInterface
```
Missing doc

dantleech commented on Event/AutoRouteCreateEvent.php:
```diff
@@ -0,0 +1,37 @@
+use Symfony\Cmf\Component\RoutingAuto\UriContext;
+use Symfony\Cmf\Component\RoutingAuto\Model\AutoRouteInterface;
+
+class AutoRouteCreateEvent extends Event
```
Add missing docs

dantleech commented on Event/AutoRouteMigrateEvent.php:
```diff
@@ -0,0 +1,38 @@
+use Symfony\Cmf\Component\RoutingAuto\UriContext;
+use Symfony\Cmf\Component\RoutingAuto\Model\AutoRouteInterface;
+
+class AutoRouteMigrateEvent extends Event
```
Docs

dantleech commented on Tests/Unit/AutoRouteManagerTest.php:
```diff
@@ -14,6 +14,9 @@
 use Symfony\Cmf\Component\RoutingAuto\UriContextCollection;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Cmf\Component\RoutingAuto\Event\UriContextEvent;
+use Symfony\Cmf\Component\RoutingAuto\RoutingAutoEvents;
```
required?

dantleech commented:
@dbu @WouterJ what do you think about this approach? We decorate the adapter with an event dispatching adapter.

Currently only 2 events are implemented, but I hope that this extension point will be all that is needed in the future. I don't thinkj we should provide events for token_providers, conflict_resolver's and defunct_route handlers as these are already pluggable.

dbu commented on RoutingAutoEvents.php:
```diff
@@ -0,0 +1,25 @@
+class RoutingAutoEvents
+{
+    /**
+     * Dispatched after the adapter has created a new auto route
```
you could mention the event class for each event as well.

dbu commented:
looks sane to me. this means that the event handlers can not change what route is created, only do additional things, right?

dantleech commented:
Yeah, I guess instantiating the Route object would be out of scope for an Event Listener (?). You can currently specify the Route FQN in the constructor of the adapter, although I guess a better solution would be a factory in the long run.

dbu commented:
one option would be to do the same as the symfony kernel and just raise events and see what the listeners put into the event object and make the default adapter an event listener too. that seems the most decoupled approach.

dantleech commented:
But then there are many methods in the adapter which are called at various times and which are more-or-less coupled to eachother (or at least copuled to the persistence implementation), so we would likely need many events and we would need to inject the dispatcher at various places, including the conflict resolver and defunct route handling classes - which would mean adding extra responsiblity to userland classes. 

But its an interesting idea, I guess the event dispatching adapter could fire the events and then the "other" adapters would just be listeners. Not sure if thats more simple though.

dbu commented:
ah right, there are way too many things. i think this would get
overcomplicated.

then what you could do would be pre and post events, and if the pre
event yields a route already, the adapter is not triggered.

WouterJ commented:
I'm missing a bit what this does. As the event is triggered after the adapter is called, none of things changed by the event listener will change anything in the database, is it? Then, what's the point of these listeners?

WouterJ commented:
Hmm, sorry. For a moment, I thought the Adapter was also persisting + flushing, but it is of course only creating the objects.

Then I'm :+1: for the current implementation, as it perfectly restricts the extension point. Can you please fix the open comments?

WouterJ commented on AutoRouteManager.php:
```diff
@@ -88,6 +103,31 @@ public function handleDefunctRoutes()
+            return $autoRoute;
+        }
+
+        $uri = $uriContext->getUri();
```
as it's overriden on the next line and not used on that line, I think this line is redundant?

dantleech replied:
Yeah, makes no sense.

WouterJ commented on AutoRouteManager.php:
```diff
@@ -88,6 +103,31 @@ public function handleDefunctRoutes()
+        if ($isSameContent) {
+            $autoRoute = $existingRoute;
+            $autoRoute->setType(AutoRouteInterface::TYPE_PRIMARY);
+            return $autoRoute;
```
there has to be an empty line before the return statement.

WouterJ commented on CHANGELOG.md:
```diff
@@ -13,10 +14,10 @@ dev-master
 * **2014-08-19**: Added validation for yaml routing auto configuration keys
 * **2014-08-10**: Replaced all instances of the term URL with URI
 * Separated this package from the RoutingAutoBundle
+
```
should be removed

WouterJ commented on CHANGELOG.md:
```diff
@@ -13,10 +14,10 @@ dev-master
                   instead of or in addition to URI strings.
 * **2015-01-29**: Added SymfonyContainerParameterTokenProvider for retrieving
                   path elements from the DI container
-
```
this should be reverted

dantleech commented on composer.json:
```diff
@@ -15,6 +15,7 @@
         "symfony-cmf/routing": "~1.2",
         "symfony/options-resolver": "~2.2",
         "symfony/config": "~2.2",
+        "symfony/event-dispatcher": "~2.2",
```
should this be a requirement?

dantleech commented:
Fixed comments and added a better test for the event dispatching adapter. Should the event dispatcher be a requirement? I guess it is actually optional.

I guess also that the follow up to this PR would be to make the RoutingAutoBundle wrap the adapters by default.

WouterJ commented:
Cool! I've made the requirement optional in https://github.com/symfony-cmf/RoutingAuto/commit/255a0f94502d15ba4bb41fe1d7b30a07edfaae4c , as you can use the component without the EventDispatcher.
